nemo/collections/nlp/data/language_modeling/megatron/megatron_batch_samplers.py

```diff
@@ -161,7 +161,8 @@ def __iter__(self):
         if len(batch) > 0 and not self.drop_last:
             # start_idx, end_idx = self.get_start_end_idx()
             indices = [batch[i] for i in range(self.data_parallel_rank, len(batch), self.data_parallel_size)]
-            # yield batch[start_idx:end_idx]
+            num_pad = self._global_batch_size // self.data_parallel_size - len(indices)
```

Comment:
This changes the behavior of batch sampler. It will impact the other datasets who uses batch sampler

Reply:
The drop_last is not set in any of the other models and actually shouldn't be supported coz its broken. The models are not equipped to handle partial batches even if the batch sampler sends the partial indices. The batch sampler also is broken coz in some scenarios when data parallel rank > number of remaining samples, it will yield an empty list. So this will break things. . Ideally we need to pad the last batch (It can happen per model, or in the sampler). Padding in sampler seems cleaner. (BTW drop last is not exposed in any of the configs except gpt3, coz thats the only model which can support drop last now) The other models ideally should just set drop last to be false)